pkg/controller/controller.go

```diff
@@ -0,0 +1,136 @@
+package controller
+
+import (
+	kcache "github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	kutil "github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// RunnableController is a controller which implements a Run loop.
+type RunnableController interface {
+	// Run starts the asynchronous controller loop.
+	Run()
+}
+
+// RetryController is a RunnableController which delegates resource
+// handling to a function and knows how to safely manage retries of a resource
+// which failed to be successfully handled.
+type RetryController struct {
+	// HandleNext is expected to process the next resource, and return the processed
+	// resource and any error that occured.
+	HandleNext func() (interface{}, error)
+
+	// RetryManager is fed the handled resource if Handle returns a Retryable
+	// error. If Handle returns no error, the RetryManager is asked to forget
+	// the resource.
+	RetryManager RetryManager
+
+	// Stop is used to shut down the controller.
+	Stop <-chan struct{}
+}
+
+// Run will call RunOnce to handle resources until a message is received on
+// the Stop channel.
+func (c *RetryController) Run() {
+	go kutil.Until(func() { c.RunOnce() }, 0, c.Stop)
+}
+
+// RunOnce calls Handle. If Handle returns a retryable error, the handled
+// resource is passed to the RetryManager. If no error is returned from
+// Handle, the RetryManager is asked to forget the processed resource.
+func (c *RetryController) RunOnce() {
+	resource, err := c.HandleNext()
+	if err != nil {
+		if _, isRetryable := err.(*RetryableError); isRetryable {
```

Comment:
Good idea... done. Was on the fence about actually defining an interface, ended up just passing a function and providing a stock implementation.